runtime/sema/check_transaction_declaration.go

```diff
@@ -248,36 +239,168 @@ func (checker *Checker) declareTransactionDeclaration(declaration *ast.Transacti
+}
+
+func (checker *Checker) transactionRoleType(declaration *ast.TransactionRoleDeclaration) *TransactionRoleType {
+	transactionRoleType := &TransactionRoleType{}
+
+	members, fields, origins := checker.compositeFieldMembersAndOrigins(
+		transactionRoleType,
+		declaration.DeclarationKind(),
+		declaration.Fields,
+	)
+
+	transactionRoleType.Members = members
+	transactionRoleType.Fields = fields
+	if checker.PositionInfo != nil {
+		checker.PositionInfo.recordMemberOrigins(transactionRoleType, origins)
+	}
+
+	if declaration.Prepare != nil {
+		parameterList := declaration.Prepare.FunctionDeclaration.ParameterList
+		transactionRoleType.PrepareParameters = checker.parameters(parameterList)
+	}
+
+	return transactionRoleType
+}
+
+func (checker *Checker) compositeFieldMembersAndOrigins(
+	containerType Type,
+	declarationKind common.DeclarationKind,
+	declarations []*ast.FieldDeclaration,
+) (
+	members *StringMemberOrderedMap,
+	fields []string,
+	origins map[string]*Origin,
+) {
 	var fieldDeclarations []ast.Declaration
 
-	fieldCount := len(declaration.Fields)
+	fieldCount := len(declarations)
 	if fieldCount > 0 {
 		fieldDeclarations = make([]ast.Declaration, fieldCount)
-		for i, field := range declaration.Fields {
+		for i, field := range declarations {
 			fieldDeclarations[i] = field
 		}
 	}
 
 	allMembers := ast.NewMembers(checker.memoryGauge, fieldDeclarations)
 
-	members, fields, origins := checker.defaultMembersAndOrigins(
+	return checker.defaultMembersAndOrigins(
 		allMembers,
-		transactionType,
+		containerType,
 		ContainerKindComposite,
-		declaration.DeclarationKind(),
+		declarationKind,
 	)
+}
 
-	transactionType.Members = members
-	transactionType.Fields = fields
-	if checker.PositionInfo != nil {
-		checker.PositionInfo.recordMemberOrigins(transactionType, origins)
+func (checker *Checker) VisitTransactionRoleDeclaration(declaration *ast.TransactionRoleDeclaration) (_ struct{}) {
+	transactionRoleType := checker.Elaboration.TransactionRoleDeclarationType(declaration)
+	// Type should have been declared in declareTransactionDeclaration
+	if transactionRoleType == nil {
+		panic(errors.NewUnreachableError())
 	}
 
-	if declaration.Prepare != nil {
-		parameterList := declaration.Prepare.FunctionDeclaration.ParameterList
-		transactionType.PrepareParameters = checker.parameters(parameterList)
+	checker.containerTypes[transactionRoleType] = struct{}{}
+	defer delete(checker.containerTypes, transactionRoleType)
+
+	fields := declaration.Fields
+	fieldMembers := orderedmap.New[MemberFieldDeclarationOrderedMap](len(fields))
+
+	for _, field := range fields {
+		fieldName := field.Identifier.Identifier
+		// Fields were previously declared in declareTransactionDeclaration
+		member, ok := transactionRoleType.Members.Get(fieldName)
+
+		if !ok {
+			panic(errors.NewUnreachableError())
+		}
+
+		fieldMembers.Set(member, field)
 	}
```

Comment:
Code is shared between `VisitTransactionRoleDeclaration` and `VisitTransactionDeclaration`; refactor to function?

Reply:
Refactored in ea613440bb9bab17c5cdcf329accb550a6fce97d. Turns out this was also duplicated in the composite declaration checking function